docs(ADR): ADR-005 Social tips contract (#93)
desmos-labs/desmos-contracts: docs/architecture/adr-005-social-tips-contract.md

# ADR 005: Social tips contract

## Changelog

- Oct 10, 2022: Initial draft;
- Oct 12, 2022: Second review;

## Status
DRAFT

## Abstract

> This ADR defines the architecture of the Social Tips contract. This contract allow sending tips to a user

Reviewer: ```suggestion
This ADR defines the architecture of the Social Tips contract. This contract allows sending tips to a user
```

using their know **handle** of a centralized application.  
The handle can be the @handle of Twitter, the user's nickname on Discord or even an e-mail address.  

> Additionally, users may want to send tips to other users that don't have a Desmos profile this contract

Reviewer: ```suggestion
Additionally, users may want to send tips to other users that haven't had a Desmos profile, this contract
```

Author: this is grammatically incorrect.

can collect those tips and allow the users to later claim it.

# Context
Desmos based applications may want to leverage on the web3 features and one of the most popular one is for sure

> the possibility to tip users. Currently, it's very easy to send token to a user knowing is address,

Reviewer: ```suggestion
the possibility to tip users. Currently, it's very easy to send tokens to a user knowing is address,
```

but it's more difficult to send tokens to a user knowing only his identity on a centralized application.

## Decision

> The idea here is to write a contract called `Social Tips`, that allows to send tokens to another user 
> through their centralized application handle. Also, if the receiver haven't linked yet his 
> centralized application handle this contract allow them to claim their tips later after having
> proved that they own such identity on the centralized application.

Reviewer: ```suggestion
The idea here is to write a contract called `Social Tips`, that allows to send tokens to another user 
through their centralized application handle. Also, if the receiver haven't linked yet his 
centralized application handle to their Desmos profile, this contract allows them to claim their tips later after having
proved that they own such identity on the centralized application.
```

## Specifications
Here below the specifications for the contract's messages:

### Messages

#### Execute
```rust
enum ExecuteMsg {
  SendTip { 
      amount: Vec<Coin>, 
      application: String,
      handle: String,
  },
  ClaimTips {},
}
```

##### SendTip
With the `SentTip` message the user can send a tip to another user though the handle of a linked centralized application.
The `MessageInfo` fields contains:
* the funds necessary to cover the fees plus the tip amount specified inside the `funds` field
* the sender identified by the `sender` field

#### ClaimTips
With the `ClaimTips` message the user can claim their pending tips that has been sent to him before proving that
owns the identities to which the tips refer.

### Query
```rust
enum QueryMsg {
  UserPendingTips { user: String }
}
```

##### UserPendingTips
With the `UserPendingTips` message the user can query the pending tips of the given `user`.
The returned tips are provided as follows:
```rust
struct QueryPendingTipsResponse {

>     pub tips: Vec<Tip>,

Reviewer 2: How's the `Tip` structure defined? Which fields does it have?

Author: it's below that struct definition at line `71`

Reviewer 2: Sorry I didn't see it! 😓 Would be useful to maybe add also a date of the tip?

Author: Ok, I add a block height field

# ADR 005: Social tips contract

## Changelog

- Oct 10, 2022: Initial draft;

## Status
DRAFT

## Abstract
This ADR defines the architecture of the Social Tips contract. This contract allow sending tips to a user
using their know **handle** of a centralized application.  

> The handle can be the @handle of Twitter, the user's nickname on Discord or even an e-mail address. 

Reviewer: I have a question, why does this contract use centralized apps handles instead of dtag?
If user want to send a tip to the receiver via contract, then he must know the dtag and address, also the application link to the profile. So, honestly, using centralized handles is redundant to me.

Reviewer: Just read the jira task description. Can you add the idea sending a user who does not have desmos profile yet?

# Context
Desmos based applications may want to leverage on the web3 features and one of the most popular one is for sure
the possibility to tip users. Currently, it's very easy to send token to a user knowing is address,
but it's more difficult to send tokens to a user knowing only his identity on a centralized application.

## Decision
The idea here is to write a contract called `Social Tips`, that allows to send tokens to another user 
through their centralized application handle. Also, if the receiver haven't linked yet his 
centralized application handle this contract allow them to claim their tips later after having
proved that they own such identity on the centralized application.

## Specifications
Here below the specifications for the contract's messages:

### Messages

#### Instantiate
```rust
struct InstantiateMsg {
  pub admin: String,
  pub subspace_id: u64,
}
```

> * The `admin` identifies the user that controls the contract;

Reviewer: What permissions does admin have in this contract?
I think this contract can make user pay the fee to the contract, then admin can claim fees to get profit from this contract, or the admin field can be removed.

* The `subspace_id` identifies the application which is deploying the contract;

#### Execute
```rust
enum ExecuteMsg {
  SendTip { 
      amount: Vec<Coin>, 

>       application: Stting,

Reviewer: ```suggestion
      application: String,
```

      handle: String,
  },
  ClaimTips {},
  UpdateAdmin { new_admin: String }
}
```

##### SendTip
With the `SentTip` message the user can send a tip to another user though the handle of a linked centralized application.
The `MessageInfo` fields contains:
* the funds necessary to cover the fees plus the tip amount specified inside the `funds` field
* the sender identified by the `sender` field

#### ClaimTips
With the `ClaimTips` message the user can claim their pending tips that has been sent to him before proving that
owns the identities to which the tips refer.

##### UpdateAdmin
With the `UpdateAdmin` message the user can update the contract's admin.
The message should make sure that the user trying to make the edit is the actual contract admin.

### Query
```rust
enum QueryMsg {
  Config {},

>   UserPendingTips { user: String }

Reviewer: ```suggestion
  UserPendingTips { application: String, user: String }
```
I think application name should be included to avoid the same name on different centralized apps.

Author: Sorry the name is misleading, `user` is the user address not it's handle
